src/topp/GNPSExport.cpp

```diff
@@ -80,296 +150,387 @@ class TOPPGNPSExport : public TOPPBase
+        generateMSMSSpectrumBins(mz_int_pairs, bin_width, ms2_block);      
 
-              double similarity_index = 5 * abs(feature.getMZ() - peptide_identification.getMZ()) +
-              abs(feature.getRT() - peptide_identification.getRT());
+        // write output
+        output_file << "BEGIN IONS" << endl;
+        output_file << "OUTPUT=" << output_type << endl;
 
-              pair<double,PeptideIdentification> first_pair = pair<double,PeptideIdentification>(similarity_index, peptide_identification);
-              pair<int,int> second_pair = pair<int,int>(map_index, spectrum_index);
+        output_file << "SCANS=" << (cons_i+1) << endl;
+        output_file << "FEATURE_ID=e_" << feature.getUniqueId() << endl;
 
-              peptides.emplace_back(first_pair,second_pair);
-            }
-          }
-          else
+        output_file << "MSLEVEL=2" << endl;
+        output_file << "CHARGE=" << to_string(charge == 0 ? 1 : charge) << "+" << endl;
+        output_file << "PEPMASS=" << feature.getMZ() << endl;
+        output_file << "FILE_INDEX=" << spec_index << endl;
+        output_file << "RTINSECONDS=" << spec.getRT() << endl;
+
+        output_file << fixed << setprecision(4);
+        for (auto ms2_iter = ms2_block.begin(); ms2_iter != ms2_block.end(); ++ms2_iter)
+        {
+          if ((int) ms2_iter->second > 0)
           {
-            should_skip_feature = true;
+            output_file << ms2_iter->first << "\t" << (int) ms2_iter->second << endl;
           }
         }
+      
+        output_file << "END IONS" << endl << endl;
       }
-
-      // peptides list of < <similarity_index, PeptideIdentification>, <map_index, feature_index> >
-
-      // with the remaining peptides left within mz/rt tol of most intense
-      if (!should_skip_feature && !peptides.empty())
+      // else output selection is 'merged_spectra' or 'full_spectra'
+      else
       {
-        // prepare peptides for output with highest mz value at top
-        sort (peptides.begin(), peptides.end(), [](const pair<pair<double,PeptideIdentification>,pair<int,int>> &a, const pair<pair<double,PeptideIdentification>,pair<int,int>> &b)
+        // discard poorer precursor spectra for 'merged_spectra' and 'full_spectra' output
+        if (pepts.size() > (unsigned long) pept_cutoff) { pepts.erase(pepts.begin()+pept_cutoff, pepts.end()); }        
+        // printf("pepts.size() %lu\n", pepts.size());
+        for (vector<pair<int,int>>::iterator pepts_iter=pepts.begin(); pepts_iter!=pepts.end(); ++pepts_iter)
         {
-          return a.first.first < b.first.first;
-        });
+          int map_index = pepts_iter->first;          
+
+          // load missing MzMLFile files + MSExperiment specs
+          if (msmaps_cached.find(map_index) == msmaps_cached.end())
+          { 
+            // printf("\t\tregistering new msmap\n");
+            MzMLFile mzmlfile;
+            mzml_files[num_msmaps_cached] = mzmlfile;
```

Comment:
no need to do this. When creating the array (or vector) it is default initialized